src/token_contract/src/main.nr

```diff
@@ -184,78 +97,58 @@ pub contract Token {
-            to,
-            from,
-        ));
-        // We don't constrain encryption of the note log in `transfer` (unlike in `transfer_in_private`) because the transfer
-        // function is only designed to be used in situations where the event is not strictly necessary (e.g. payment to
-        // another person where the payment is considered to be successful when the other party successfully decrypts a
-        // note).
-        Transfer { from, to, amount }.emit(encode_and_encrypt_event_unconstrained(
+        storage.private_balances.at(to).add(to, amount).emit(encode_and_encrypt_note(
             &mut context,
             to,
             from,
         ));
     }
 
-    #[contract_library_method]
-    fn subtract_balance(
-        context: &mut PrivateContext,
-        storage: Storage<&mut PrivateContext>,
-        account: AztecAddress,
+    // Transfers token `amount` from public balance of message sender to a private balance of `to`.
+    #[private]
+    fn transfer_public_to_private(
+        from: AztecAddress,
+        to: AztecAddress,
         amount: U128,
-        max_notes: u32,
-    ) -> U128 {
-        let subtracted = storage.balances.at(account).try_sub(amount, max_notes);
-        // Failing to subtract any amount means that the owner was unable to produce more notes that could be nullified.
-        // We could in some cases fail early inside try_sub if we detected that fewer notes than the maximum were
-        // returned and we were still unable to reach the target amount, but that'd make the code more complicated, and
-        // optimizing for the failure scenario is not as important.
-        assert(subtracted > U128::zero(), "Balance too low");
-        if subtracted >= amount {
-            // We have achieved our goal of nullifying notes that add up to more than amount, so we return the change
-            subtracted - amount
-        } else {
-            // try_sub failed to nullify enough notes to reach the target amount, so we compute the amount remaining
-            // and try again.
-            let remaining = amount - subtracted;
-            compute_recurse_subtract_balance_call(*context, account, remaining).call(context)
-        }
-    }
+        nonce: Field,
+    ) {
+        _validate_from_private(from, nonce, &mut context);
 
-    // TODO(#7729): apply no_predicates to the contract interface method directly instead of having to use a wrapper
-    // like we do here.
-    #[no_predicates]
-    #[contract_library_method]
-    fn compute_recurse_subtract_balance_call(
-        context: PrivateContext,
-        account: AztecAddress,
-        remaining: U128,
-    ) -> PrivateCallInterface<25, U128> {
-        Token::at(context.this_address())._recurse_subtract_balance(account, remaining)
+        // `from` is the owner of the public balance from which we'll subtract the `amount`.
+        let from = context.msg_sender();
+        let token = Token::at(context.this_address());
```

Comment:
we sometimes declare it like this: `let token = Token::at(...)` and sometimes use straight `Token::at(...)`, let's standarize